### DISCUSSION TITLE
About unexpected char sequences into binary data

### DISCUSSION THREAD

**Discussion by saoirse-iontach** [gildas-lormeau/Polyglot-HTML-ZIP-PNG#2] [General]
Binary data are wrapped into `<!-- ... -->` comments.
But data may contains a `-->` char sequence!
Then we should reopen comment, but the parser doesn't allow that.

Instead of comment, we could wrap into script element,
Then handle closing with a custom `src` script.

```
<script>function onclose(){
  // if (true) /* then ends on EOF with a noop document write */
  // if (count--) /* pre recorded value */
  // if (!...endsWith('${win_1252_data.slice(-42)}')) /* pre recorded value */
  if (!document.currentScript.textContent.endsWith('===EOD===')) /* unseen custom marker */
  {
    document.write('<script src="data:,onclose()">');
  }
}</script>

<script src="data:,onclose()">Wrapped binary data....
...with unexpected</script>and continuation...
...with another unexpected</script probably illformed...>and...
...and finally ===EOD===</script>
```

Note that we have to restore original `</script...>` when joining all script text contents.

-----
So in build process:
 - If data doesn't contains `-->`, then use comment
 - else if data doesn't contains `</script`, then use (no executable) script element
 - else do complex build:
   - record all `</script...>` end tag
   - use script element with close handler
   - use appropriated post-processing to join all script text contents and recorded end tags.

**Comment by gildas-lormeau** [comment]
Actually, this is already implemented in SingleFile, here are the list of alternate tags which can be used when the binary data contains the closing tag: https://github.com/gildas-lormeau/single-file-core/blob/main/processors/compression/compression.js#L44-L49.

You can see it in action in the presentation itself (https://github.com/gildas-lormeau/Polyglot-HTML-ZIP-PNG/raw/main/en-EN/dist/presentation-polyglot-png-zip-html_en-EN.html) which uses good old `<xmp>` and `</xmp>` tags for the ZIP data.

**Comment by saoirse-iontach** [comment]
Effectively, usable tags for *RawText* (and *Script Data*) are:
 - script and style (with unknow type)
 - noscript, noframes, noembed,
 - xpm
 - iframe
 - plaintext (witch runs to EOF!)

About plaintext, there is no end tag. It may be very interesting!

About end tag rexexp, they should include `\b[^>]*` &nbsp; <sub>*(=> `/<\/tag\b[^>]*>/i` )*</sub>

Thank you for all your help :-)

**Reply by gildas-lormeau** [reply]
Thank you for the info! I'll try to add the missing tags in SingleFile.
